src/plugins/typechecker/README.md

```diff
@@ -0,0 +1,56 @@
+- infos = Information about the typechecker plugin is in keys below
+- infos/author = e1528532 <e1528532@libelektra.org>
+- infos/licence = BSD
+- infos/needs = 
+- infos/provides = typechecker
+- infos/recommends =
+- infos/placements = postgetstorage presetstorage
+- infos/status = maintained experimental global
+- infos/metadata =
+- infos/description = a plugin which typechecks configuration specifications
+
+## Introduction
+
+A plugin which typechecks specifications before setting keys and after getting keys from
+a mounted specification. 
+
+## Usage
+
+In order to use the type checker, mount a configuration specification along with this
+plugin. 
+
+`kdb mount <specification> spec/<path> <storage plugin to read the specification> typechecker`
```

Comment:
It did not work for me :cry: 

For me the execution hangs:
```
kdb mount ~e/src/plugins/typechecker/simpleTestSpecification.ini spec/examples ini typechecker
kdb ls spec/examples
# typechecker: src/Control/Monad/Ghc.hs:53:10-55: No instance nor default method for class operation generalBracket
# <does not terminate>
```
(get or ls does not make a difference)

Please make your "kdb mount" concrete so that people can actually try it out (with some "kdb set" which produce type errors), ideally checked with shell recorder.

Furthermore the infos in this README are not really helpful how the typechecking works. Most likely we need a separate tutorial for that.

Reply:
Interesting, i can't reproduce this in my debian stretch. I knew i tried it there once 🤔 
I have the same ghc, cabal and c2hs versions as you do. The error seems to happen inside some ghc-internal library which is odd. The internal libraries are shipped with ghc so they should be the same as the ones i have. Lets look into that when we meetup on thu.

I have identified an issue when i renamed the keys though, that should be fixed and i get some useable results in my debian again.

I agree the docu is very minimalistic, i want to avoid the effort of writing too much docu yet while the underlying system might still change a bit so my intention was to do that when the formalization is fixed and final. So for demonstration purposes i have only included this simpleTestSpecification.ini to show the basic principle.

Some shell recorder tests seem to be a very good idea for the future, but as i said i wanted to avoid too much work on that while the underlying specification is still not final. They are definitely useful for both demonstration and testing though!

Reply:
Now the compilation fails [0]. I think a protoype with tutorial/docu is essential in order to evaluate if the type checker is practical to work with. Unittests and the proof can be delayed.

[0]:
```
/home/markus/Projekte/Elektra/current/src/plugins/typechecker/Elektra/Typechecker.hs:28:16: error:
    • Couldn't match expected type ‘KeySet -> IO [TypeSpecification]’
                  with actual type ‘IO [TypeSpecification]’
    • The function ‘parseTypeSpecifications’
      is applied to two arguments,
      but its type ‘KeySet -> IO [TypeSpecification]’ has only one
      In a stmt of a 'do' block:
        typeSpecs <- parseTypeSpecifications k ks
      In the second argument of ‘($)’, namely
        ‘do { logDebug "Parse Specifications now.";
              typeSpecs <- parseTypeSpecifications k ks;
              logDebugT
              $ "Got " ++ show (length typeSpecs) ++ " function specifications";
              mapM_ (logDebugT . show) typeSpecs;
              .... }’

/home/markus/Projekte/Elektra/current/src/plugins/typechecker/Elektra/Typechecker.hs:28:40: error:
    • Couldn't match expected type ‘KeySet’ with actual type ‘Key’
    • In the first argument of ‘parseTypeSpecifications’, namely ‘k’
      In a stmt of a 'do' block:
        typeSpecs <- parseTypeSpecifications k ks
      In the second argument of ‘($)’, namely
        ‘do { logDebug "Parse Specifications now.";
              typeSpecs <- parseTypeSpecifications k ks;
              logDebugT
              $ "Got " ++ show (length typeSpecs) ++ " function specifications";
              mapM_ (logDebugT . show) typeSpecs;
              .... }’

/home/markus/Projekte/Elektra/current/src/plugins/typechecker/Elektra/Typechecker.hs:32:15: error:
    • Couldn't match expected type ‘KeySet -> IO [KeySpecification]’
                  with actual type ‘IO [KeySpecification]’
    • The function ‘parseKeySpecifications’
      is applied to two arguments,
      but its type ‘KeySet -> IO [KeySpecification]’ has only one
      In a stmt of a 'do' block: keySpecs <- parseKeySpecifications k ks
      In the second argument of ‘($)’, namely
        ‘do { logDebug "Parse Specifications now.";
              typeSpecs <- parseTypeSpecifications k ks;
              logDebugT
              $ "Got " ++ show (length typeSpecs) ++ " function specifications";
              mapM_ (logDebugT . show) typeSpecs;
              .... }’

/home/markus/Projekte/Elektra/current/src/plugins/typechecker/Elektra/Typechecker.hs:32:38: error:
    • Couldn't match expected type ‘KeySet’ with actual type ‘Key’
    • In the first argument of ‘parseKeySpecifications’, namely ‘k’
      In a stmt of a 'do' block: keySpecs <- parseKeySpecifications k ks
      In the second argument of ‘($)’, namely
        ‘do { logDebug "Parse Specifications now.";
              typeSpecs <- parseTypeSpecifications k ks;
              logDebugT
              $ "Got " ++ show (length typeSpecs) ++ " function specifications";
              mapM_ (logDebugT . show) typeSpecs;
              .... }’
```